sklearn/datasets/_openml.py

```diff
@@ -105,22 +106,20 @@ def is_gzip_encoded(_fsrc):
         return fsrc
 
     local_path = _get_local_path(openml_path, data_home)
+    dir_name, file_name = os.path.split(local_path)
     if not os.path.exists(local_path):
+        os.makedirs(dir_name, exist_ok=True)
         try:
-            os.makedirs(os.path.dirname(local_path))
-        except OSError:
-            # potentially, the directory has been created already
-            pass
-
-        try:
-            with closing(urlopen(req)) as fsrc:
-                opener: Callable
-                if is_gzip_encoded(fsrc):
-                    opener = open
-                else:
-                    opener = gzip.GzipFile
-                with opener(local_path, "wb") as fdst:
-                    shutil.copyfileobj(fsrc, fdst)
+            with TemporaryDirectory(dir=dir_name) as tmpdir:
+                with closing(urlopen(req)) as fsrc:
+                    opener: Callable
+                    if is_gzip_encoded(fsrc):
+                        opener = open
+                    else:
+                        opener = gzip.GzipFile
+                    with opener(os.path.join(tmpdir, file_name), "wb") as fdst:
+                        shutil.copyfileobj(fsrc, fdst)
+                shutil.move(fdst.name, local_path)
```

Comment:
I think this is mostly safe as long as the temporary directory and the destination are in the same filesystem. According to [shutil.move](https://docs.python.org/3/library/shutil.html#shutil.move) docs, `shutil.move` is the same as `os.rename` on the same filesystem.

If there is actually a copy, I can see issues such as:

- If two threads try to copy to the same location at the same time, I think the file can end up corrupted.
- If one thread is in the process of copying over the file, another thread will see that file exist and attempt to read it, which would fail since the file is not fully copied yet.

I do not know a great way forward. Using `os.rename` directly may break use cases where the temporary directory is not on the same filesystem as the cache. And copying itself may lead to errors as stated above.

Reply:
My original idea was to use a naming scheme based on process id or thread Id and create a soft symlik instead of renaming. If same process wants to clean cache remove the original and symlink If another process tries to dow write data create a new file based on its id and for reading it can use the original symlink. What do you think @thomasjpfan ?

Reply:
I think the tempfile is guaranteed to be in the same filesystem and therefore the renaming to be atomic because we create the tempfolder as a subfolder.

So I think the current code is fine. Maybe @siavrez you can add an inline comment to explain this.

Reply:
> I think the tempfile is guaranteed to be in the same filesystem and therefore the renaming to be atomic because we create the tempfolder as a subfolder.

Ah I see it now. Thank you for the explanation.